### PR TITLE
test: enable windows unit tests with `csi-proxy.exe` 

### PR DESCRIPTION
**Pull request by mayankshah1607**
Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

**What type of PR is this?**
/kind test

**What this PR does / why we need it**:
test: enable windows unit tests with `csi-proxy.exe` 

**Which issue(s) this PR fixes**:
https://github.com/kubernetes-csi/csi-driver-smb/issues/121
https://github.com/kubernetes-csi/csi-driver-smb/issues/109

Fixes #

**Requirements**:
- [x] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [x] includes documentation
- [x] adds unit tests
- [ ] tested upgrade from previous version

**Special notes for your reviewer**:

**Release note**:
```
none
```

### CODE REVIEW

File: pkg/smb/nodeserver_test.go [hunk 1]
```diff
@@ -34,6 +34,19 @@ import (
...
+		fmt.Sprintf("volume(vol_1##) mount \"test_source\" on %#v"+
+			" failed with smb mapping failed with error: rpc error: code = "+
+			"Unknown desc = NewSmbGlobalMapping failed. output: \"New-SmbGlobalMapping"+
+			" : Access is denied. \\r\\nAt line:1 char:190\\r\\n+ ... ser, "+
+			"$PWord;New-SmbGlobalMapping -RemotePath $Env:smbremotepath -Cred ...\\r\\n+"+
+			"                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~"+
+			"\\r\\n    + CategoryInfo          : PermissionDenied: (MSFT_SmbGlobalMapping:ROOT/Microsoft/...mbGlobalMapping)"+
+			" [New-SmbGlobalMapping], Cim \\r\\n   Exception\\r\\n    + FullyQualifiedErrorId : "+
+			"Windows System Error 5,New-SmbGlobalMapping\\r\\n \\r\\n\", err: exit status 1", file))
+}
```
Comment (mayankshah1607) [inline pkg/smb/nodeserver_test.go:48]: @andyzhangx These error messages seem pretty flaky. I'm getting a slightly different message on my local Windows machine, but Github Action gives a different one. Any ideas about how we can go about asserting these errors?

Comment (andyzhangx) [reply pkg/smb/nodeserver_test.go:49]: use error msg matching should be enough, e.g. match `Access is denied`

Comment (mayankshah1607) [reply pkg/smb/nodeserver_test.go:49]: We're not getting `Access is Denied` error on Github Actions. Instead we're getting `The parameter is incorrect`

```
Unexpected error: rpc error: code = Internal desc = volume(vol_1##) mount "test_source" on "D:\\a\\csi-driver-smb\\csi-driver-smb\\pkg\\smb\\source_test" failed with smb mapping failed with error: rpc error: code = Unknown desc = NewSmbGlobalMapping failed. output: "New-SmbGlobalMapping : The parameter is incorrect. \r\nAt line:1 char:190\r\n+ ... ser, $PWord;New-SmbGlobalMapping -RemotePath $Env:smbremotepath -Cred ...\r\n+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\r\n    + CategoryInfo          : InvalidArgument: (MSFT_SmbGlobalMapping:ROOT/Microsoft/...mbGlobalMapping) [New-SmbGloba \r\n   lMapping], CimException\r\n    + FullyQualifiedErrorId : Windows System Error 87,New-SmbGlobalMapping\r\n \r\n", err: exit status 1         
```

Comment (andyzhangx) [reply pkg/smb/nodeserver_test.go]: it failed here: https://github.com/kubernetes-csi/csi-proxy/blob/25491d0543e53381aab69ee5e5fa114182787b43/internal/os/smb/api.go#L68
could be different error description on Windows, you can try above command to verify.

File: pkg/smb/nodeserver_test.go [hunk 2]
```diff
@@ -58,6 +71,12 @@ func TestNodeStageVolume(t *testing.T) {
 		desc        string
 		req         csi.NodeStageVolumeRequest
 		expectedErr testutil.TestError
+
+		// use the below fields only for flaky Windows error messages
+
+		// if true, use substring to match errors
+		isWindowsErrorFlaky bool
+		requiredErrorSubstr string
```
Comment (mayankshah1607) [inline pkg/smb/nodeserver_test.go:79]: @andyzhangx Only for `TestNodeStageVolume`, I have added functionality to check for error messages using substring as some parts of the error messages may change, and are a bit flaky. This way the base error message is verified on Windows to tackle the flakiness.

File: pkg/smb/nodeserver_test.go [hunk 3]
```diff
@@ -277,135 +320,161 @@ func TestNodePublishVolume(t *testing.T) {
...
 		{
-			desc:        "[Error] Unmount error mocked by IsLikelyNotMountPoint",
-			req:         csi.NodeUnpublishVolumeRequest{TargetPath: errorTarget, VolumeId: "vol_1"},
-			expectedErr: status.Error(codes.Internal, "failed to unmount target \"./error_is_likely_target\": fake IsLikelyNotMountPoint: fake error"),
+			desc: "[Error] Unmount error mocked by IsLikelyNotMountPoint",
+			req:  csi.NodeUnpublishVolumeRequest{TargetPath: errorTarget, VolumeId: "vol_1"},
+			// todo: This test does not return any error on windows
+			// Once the issue is figured out, we'll remove this field
+			skipOnWindows: true,
+			expectedErr: testutil.TestError{
```
Comment (mayankshah1607) [inline pkg/smb/nodeserver_test.go:383]: In some cases, CSI proxy does not return error on Windows. Only these specific test cases are skipped on Windows for now, and the remaining test cases are executed. I think we may have to change the assertion logic a bit around here as mount behavior is different on Windows and Linux. I will open a follow-up PR to enable this test case.

**Comment by mayankshah1607** [comment]
@andyzhangx Left some comments above regarding mount behavior on Linux and Windows. If this approach looks good, we can merge this and I can enable the same approach for azuredisk and azurefile repos :)

File: pkg/smb/nodeserver_test.go [hunk 4]
```diff
@@ -58,6 +71,12 @@ func TestNodeStageVolume(t *testing.T) {
 		desc        string
 		req         csi.NodeStageVolumeRequest
 		expectedErr testutil.TestError
+
+		// use the below fields only for flaky Windows error messages
+
+		// if true, use substring to match errors
+		isWindowsErrorFlaky bool
```
Comment (andyzhangx) [inline pkg/smb/nodeserver_test.go:78]: what about only use `requiredErrorSubstr`? it seems `isWindowsErrorFlaky` could be removed

**Comment by mayankshah1607** [comment]
Thanks @andyzhangx , I have made changes as per above comments

**Comment by andyzhangx** [comment]
could you also squash all commits? thanks.

**Comment by mayankshah1607** [comment]
@andyzhangx Thanks, I have rebased and squashed the commits.

File: pkg/smb/nodeserver_test.go [hunk 5]
```diff
@@ -34,6 +36,17 @@ import (
 	"k8s.io/utils/mount"
 )
 
+func matchFlakyWindowsError(mainError error, substr string) bool {
+	var errorMessage string
```
Comment (andyzhangx) [inline pkg/smb/nodeserver_test.go:40]: nit: 
```
errorMessage := ""
if mainError != nil {
   errorMessage = mainError.Error()
}
```

**Review by andyzhangx — APPROVED** [review]
/lgtm
/approve

**Pull request merged**